### PR TITLE
Improvements to Deneb `store` upon review

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Part of #4676

## Proposed Changes

- [x] Remove unnecessary lazy_static for `MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUESTS`.
- [x] Fix bugs in blob pruning (`blob_sidecar_exists` was buggy and caused pruning to always no-op).
- [x] Clarify code comments and logs in blob pruning. This just felt natural as I was poking around and improving my understanding.
- [x] Remove unnecessary `BeaconBlobOrphan` column.
- [x] Add a database schema migration with no-op upgrade, and a no-op downgrade that only suceeeds _before_ Deneb.
- [x] Test for blob pruning happy case.
- [x] Test for blob pruning without Deneb (should return early).
- [x] Test for blob pruning without finalization (should return early).
- [x] Test for pruning wrt blob margin config

### CODE REVIEW

File: consensus/types/src/consts.rs [hunk 1]
```diff
@@ -32,9 +32,9 @@ pub mod deneb {
             "52435875175126190479447740508185965837690552500527637822603658699938581184513"
```
Comment (jimmygchen) [inline consensus/types/src/consts.rs]: just discussed with @michaelsproul and might be worth moving this to a const as well

File: beacon_node/beacon_chain/src/schema_change/migration_schema_v18.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        .map(|slot| slot.epoch(T::EthSpec::slots_per_epoch()))
+        .ok_or(Error::SlotClockUnavailableForMigration)
+}
+
+pub fn upgrade_to_v18<T: BeaconChainTypes>(
+    db: Arc<HotColdDB<T::EthSpec, T::HotStore, T::ColdStore>>,
+    log: Logger,
+) -> Result<Vec<KeyValueStoreOp>, Error> {
+    // No-op, even if Deneb has already occured. The database is probably borked in this case, but
+    // *maybe* the fork recovery will revert the minority fork and suceed.
```
Comment (realbigsean) [inline beacon_node/beacon_chain/src/schema_change/migration_schema_v18.rs]:
```suggestion
    // No-op, even if Deneb has already occurred. The database is probably borked in this case, but
    // *maybe* the fork recovery will revert the minority fork and succeed.
```

File: beacon_node/beacon_chain/src/historical_blocks.rs [hunk 3]
```diff
@@ -206,15 +209,32 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         self.store.hot_db.do_atomically(hot_batch)?;
         self.store.cold_db.do_atomically(cold_batch)?;
 
+        let mut anchor_and_blob_batch = Vec::with_capacity(2);
+
+        // Update the blob info.
+        if let Some(oldest_blob_slot) = new_oldest_blob_slot {
```
Comment (realbigsean) [inline beacon_node/beacon_chain/src/historical_blocks.rs]: should we skip this if `new_oldest_blob_slot` is unchanged?

File: beacon_node/store/src/hot_cold_store.rs [hunk 4]
```diff
@@ -2156,19 +2182,17 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
                 Err(e) => {
                     warn!(
                         self.log,
-                        "Stopping blobs sidecar pruning early";
+                        "Stopping blob pruning early";
```
Comment (realbigsean) [inline beacon_node/store/src/hot_cold_store.rs:2182]: love that we're pruning the name `sidecar` from our codebase

File: beacon_node/store/src/hot_cold_store.rs [hunk 5]
```diff
@@ -2047,107 +2044,136 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
...
         data_availability_boundary: Epoch,
     ) -> Result<(), Error> {
-        let deneb_fork = match self.spec.deneb_fork_epoch {
+        let _deneb_fork_epoch = match self.spec.deneb_fork_epoch {
             Some(epoch) => epoch,
             None => {
                 debug!(self.log, "Deneb fork is disabled");
                 return Ok(());
             }
         };
```
Comment (realbigsean) [inline beacon_node/store/src/hot_cold_store.rs]:
```suggestion
        if self.spec.deneb_fork_epoch.is_none() {
            debug!(self.log, "Deneb fork is disabled");
            return Ok(());
        }
```

File: beacon_node/store/src/hot_cold_store.rs [hunk 6]
```diff
@@ -2156,19 +2182,17 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
...
                     );
                     break;
                 }
             };
 
-            if Some(block_root) != last_pruned_block_root
-                && self.blobs_sidecar_exists(&block_root)?
-            {
+            if Some(block_root) != last_pruned_block_root && self.blobs_exist(&block_root)? {
                 debug!(
```
Comment (realbigsean) [inline beacon_node/store/src/hot_cold_store.rs]: should this be a trace if we're emitting it up to once per block on a prune? Could be pretty spammy

File: beacon_node/beacon_chain/src/canonical_head.rs [hunk 7]
```diff
@@ -1064,6 +1058,12 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             self.head_tracker.clone(),
         )?;
 
+        // Prune blobs in the background.
```
Comment (realbigsean) [inline beacon_node/beacon_chain/src/canonical_head.rs:1061]: Just noting: moving this to `after_finalization` definitely make sense with how we're currently pruning. But we'll have to move it if we want to extend pruning to work for unfinalized blobs in a future PR

**Review by realbigsean — CHANGES_REQUESTED** [review]
Awesome! Looks a lot simpler and the new tests are great

File: beacon_node/store/src/hot_cold_store.rs (hunk 4 above)
Comment (michaelsproul) [reply beacon_node/store/src/hot_cold_store.rs:2182]: I was getting irked by the weird pluralisation of "blobs sidecars", so I just nuked it

**Review by realbigsean — APPROVED** [review]
Looks like all comments are addressed, I'm going to give this a whirl on devnet 8

**Comment by realbigsean** [comment]
looks like it works on devnet 8

**Pull request merged**